Add Histogram description to DataModel Specification (#1664)
open-telemetry/opentelemetry-specification: specification/metrics/datamodel.md

### Histogram

> [Histgram](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L225)

Reviewer: ```suggestion
[Histogram](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L225)
```

metric data points convey a population of recorded data points in a compressed
format. A histogram bundles a set of events into divided populations with an
overall event count and aggregate sum for all events.

![Delta Histogram](img/model-delta-histogram.png)

Histograms consist of the following:

- An *Aggregation Temporality* of delta or cumulative.
- A set of data points, each containing:
  - An independent set of Attribute name-value pairs.
  - A time window (of `(start, end]`) time for which the Histogram was bundled.
    - The time interval is inclusive of the end time.

>     - Times are specified in Value is UNIX Epoch time in nanoseconds since
>       `00:00:00 UTC on 1 January 1970`

Reviewer: I can't quite make sense of this sentence as written. I tried for a rephrase but this might not be what you had in mind either.

```suggestion
    - Time values are specified as nanoseconds since the UNIX Epoch (00:00:00 UTC on 1 January 1970).
```

  - A count (`count`) of the total population of points in the histogram.
  - A sum (`sum`) of all the values in the histogram.

>   - (optional) A series of buckets with:

Reviewer: In future, do we intend for this to also support the automatic exponential bucketing scheme that's being designed in OM/Prom? Or will we model that as a distinct aggregation type?

Author: See the PR description.  Yes we plan to add that, and yes this will get updated when we do.

    - Explicit boundary values.  These values denote the lower and upper bounds
      for buckets and whether not a given observation would be recorded in this
      bucket.
    - A count of the number of observations that fell within this bucket.
  - (optional) a set of examplars (see [Exemplars](#exemplars)).  

Like Sums, Histograms also define an aggregation temporality.  The picture above
denotes Delta temporality where accumulated events are dropped after reporting

> and a new aggregation occurs. Cumulative, on the other hand, contiues to

Reviewer 2: ```suggestion
and a new aggregation occurs. Cumulative, on the other hand, continues to
```

aggregate events until a reset.

#### Exemplars
  
An exemplar is a recorded value that associates OpenTelemetry context to

> a metric event within a Histogram. It allows users to link Trace signals w/

Reviewer: ```suggestion
a metric event within a Histogram. One use case is to allow users to link Trace signals w/
```

### Histogram

[Histogram](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L225)

> metric data points convey a population of recorded data points in a compressed

Reviewer 3: ```suggestion
metric data points convey a population of recorded measurements in a compressed
```
```suggestion
metric data points convey a population of recorded data points in a compressed
```

format. A histogram bundles a set of events into divided populations with an
overall event count and aggregate sum for all events.

![Delta Histogram](img/model-delta-histogram.png)

Histograms consist of the following:

- An *Aggregation Temporality* of delta or cumulative.
- A set of data points, each containing:
  - An independent set of Attribute name-value pairs.
  - A time window (of `(start, end]`) time for which the Histogram was bundled.
    - The time interval is inclusive of the end time.
    - Time values are specified as nanoseconds since the UNIX Epoch
      (00:00:00 UTC on 1 January 1970).
  - A count (`count`) of the total population of points in the histogram.
  - A sum (`sum`) of all the values in the histogram.
  - (optional) A series of buckets with:
    - Explicit boundary values.  These values denote the lower and upper bounds
      for buckets and whether not a given observation would be recorded in this
      bucket.
    - A count of the number of observations that fell within this bucket.

>   - (optional) a set of examplars (see [Exemplars](#exemplars)).  

Reviewer 3: The other types have them already. Until we give some kind of sample-count or weight to the exemplars, there's not really a difference which point kind they are attached to. I have in mind adding some specs on this after https://github.com/open-telemetry/oteps/pull/148 merges.

>   - (optional) a set of examplars (see [Exemplars](#exemplars)).  

Author: This is my bad, apparently I can't search a proto file.  Missed the existence of exemplar on NumberDataPoint.

I've update the whole PR to denote Exemplar and call out its meaning, PTAL.

Like Sums, Histograms also define an aggregation temporality.  The picture above

> denotes Delta temporality where accumulated events are dropped after reporting

Reviewer 3: ```suggestion
denotes Delta temporality where accumulated event counts are reset to zero after reporting
```
```suggestion
denotes Delta temporality where accumulated events are dropped after reporting
```

and a new aggregation occurs. Cumulative, on the other hand, continues to

> aggregate events until a reset.

Reviewer 3: ```suggestion
aggregate events, resetting with the use of a new start time.
```
```suggestion
aggregate events until a reset.
```
